app/src/main/java/com/github/sdpsharelook/onlinePictures/GoogleImageApi.kt

```diff
@@ -0,0 +1,87 @@
+package com.github.sdpsharelook.onlinePictures
+
+import android.os.Build
+import com.github.sdpsharelook.language.Language
+import org.json.JSONObject
+import org.json.JSONTokener
+import java.io.BufferedReader
+import java.io.InputStreamReader
+import java.net.HttpURLConnection
+import java.net.URL
+
+// custom search api :
+
+
+/**
+ * Implements the GoogleImageApi.
+ * Try it here https://developers.google.com/custom-search/v1/reference/rest/v1/cse/list
+ */
+class GoogleImageApi {
+    companion object {
+        private val api_endpoint = "https://www.googleapis.com/customsearch/v1"
+
+        // key = AIzaSyCZAkH_vbNiB8QeyFs9yFp6AUR2eyW0uus
+        private val api_key = "AIzaSyCZAkH_vbNiB8QeyFs9yFp6AUR2eyW0uus"
+
+        // custom search engine : 82269d15153934931 (https://cse.google.com/cse?cx=82269d15153934931)
+        private val cx = "82269d15153934931"
+        private val validCharsURL =
+            (('A'..'Z') + ('a'..'z') + ('0'..'9')).toSet() + setOf('%', '~', '_', '.')
+
+        private fun sanitizeURL(url: String) =
+            url.map {
+                if (validCharsURL.contains(it)) it
+                else "%${it.code.toString(16)}"
+            }.joinToString("")
+
+        /**
+         * must be launched on Dispatchers.IO
```

Comment:
there is a way to do that without asking the outside function to do it: wrap the function body in a withContext(Dispatchers.IO) call, but that can be done later